pdns/dnsrecords.cc

```diff
@@ -489,6 +489,242 @@ string EUI64RecordContent::getZoneRepresentation(bool noDot) const
+    for (int i=0; i < 4; i++) {
+      if (i < ret->d_afdlength)
+        pr.xfr8BitInt(ret->d_ip4[i]);
+    }
+  } else if (ret->d_family == APL_FAMILY_IPV6) {
+    if (ret->d_afdlength > 16) {
+      throw MOADNSException("Invalid IP length for IPv6 APL");
+    }
+    bzero(ret->d_ip6, sizeof(ret->d_ip6));
+    for (int i=0; i< 16; i++) {
+      if (i < ret->d_afdlength)
+        pr.xfr8BitInt(ret->d_ip6[i]);
+    }
+  } else
+    throw MOADNSException("Unknown family for APL record");
+
+  return ret;
+}
+
+// Parse backend record
+std::shared_ptr<DNSRecordContent> APLRecordContent::make(const string& zone) {
+  string record;
+
+  auto ret=std::make_shared<APLRecordContent>();
+
+  // Strip the optional leading ! (negate)
+  if (zone[0] == '!') {
+    ret->d_n = 1;
+    record = zone.substr(1, zone.length()-1);
+  } else {
+    ret->d_n = 0;
+    record = zone;
+  }
+
+  if (record.find("1:", 0) == 0) { // IPv4
+    unsigned int prefix;
+    uint32_t v4ip;
+    string ipstr, subnetstr;
+    int subnet_pos;
+    ComboAddress ca;
+    int bytes;
+
+    ret->d_family = APL_FAMILY_IPV4;
+
+    // Find netmask
+     subnet_pos = record.rfind("/");
+
+    if (subnet_pos < 0) {
+      throw MOADNSException("Asked to decode '"+zone+"' as an APL record, but missing subnet mask");
+    }
+
+    // Read IPv4 string into a ComboAddress
+    ipstr = record.substr(2, subnet_pos - 2);
+    subnetstr = record.substr(subnet_pos + 1, record.length() - subnet_pos - 1);
+    ca = makeComboAddress(ipstr);
+    v4ip = ntohl(ca.sin4.sin_addr.s_addr);
+    if (ca.sin4.sin_family != AF_INET) { // ComboAddress will match v6 IPs, which is not valid here
+      throw MOADNSException("Asked to decode '"+zone+"' as an APL record, but found invalid v4 IP address");
+    }
+    // Section 4.1 of RFC 3123 (don't send trailing "0" bytes)
+    // Copy data; using array of bytes since we might end up truncating them in the packet
+    bzero(ret->d_ip4, sizeof(ret->d_ip4));
+    for (int i=0; i<4; i++) {
+      ret->d_ip4[3-i] = (v4ip & 255);
+      v4ip = v4ip >> 8;
+    }
+    // Remove trailing "0" bytes from packet and calculate length
+    bytes  = 4; // Start by assuming we'll send 4 bytes
+    v4ip = ntohl(ca.sin4.sin_addr.s_addr);
+    for (int i=0; i<4; i++) {
+      if ((v4ip & 255) == 0) {
+        // trailing 0 byte, reduce length
+        bytes--;
+      } else
+      {
+        // Found non-0 byte, stop trimming
+        break;
+      }
+      v4ip = v4ip >> 8;
+    }
+
+    // Prefix length
+    if (sscanf(subnetstr.c_str(), "%u", &prefix) == 1) {
+      ret->d_prefix=prefix;
+    } else
+          throw MOADNSException("Asked to decode '"+zone+"' as an IPv4 APL record, but found invalid prefix string "+subnetstr);
+    if (prefix > 32)
+      throw MOADNSException("Asked to decode '"+zone+"' as an IPv4 APL record, but found invalid prefix value "+std::to_string(prefix));
+    ret->d_afdlength = bytes;
+
+  } else if (record.find("2:", 0) == 0) { // IPv6
```

Comment:
there is a Netmask class that can be used to validate & manipulate netmasks with more confidence

Reply:
Thanks for the pointer. Switched the code to using the Netmask class in https://github.com/PowerDNS/pdns/pull/8638/commits/338b1fdc7be409c52b4daff7aa7a0260e597e4a2